cmd/stellar-rpc/internal/ingest/backfill.go

```diff
@@ -0,0 +1,278 @@
+package ingest
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/stellar/go-stellar-sdk/ingest/ledgerbackend"
+	"github.com/stellar/go-stellar-sdk/support/datastore"
+	supportlog "github.com/stellar/go-stellar-sdk/support/log"
+	"github.com/stellar/go-stellar-sdk/xdr"
+
+	"github.com/stellar/stellar-rpc/cmd/stellar-rpc/internal/config"
+	"github.com/stellar/stellar-rpc/cmd/stellar-rpc/internal/db"
+)
+
+const (
+	// OneDayOfLedgers is (roughly) a 24 hour window of ledgers.
+	OneDayOfLedgers   = config.OneDayOfLedgers
+	SevenDayOfLedgers = config.OneDayOfLedgers * 7
+	// Number of ledgers to read/write at a time during backfill
+	ChunkSize uint32 = OneDayOfLedgers / 4 // 6 hours. Takes X minutes to process
+)
+
+// This function backfills the local database with n ledgers from the datastore
+// It is called by daemon.go if cfg.Backfill is true
+func RunBackfill(cfg *config.Config, logger *supportlog.Entry, localDbConn *db.DB, localDbRW db.ReadWriter, dsInfo DatastoreInfo) error {
+	logger.Infof("Beginning backfill process")
+	var (
+		ctx              context.Context = context.Background()
+		currentTipLedger uint32
+		dbIsEmpty        bool = false
+
+		nBackfill     uint32          = cfg.HistoryRetentionWindow
+		localDbPath   string          = cfg.SQLiteDBPath
+		localDbReader db.LedgerReader = db.NewLedgerReader(localDbConn)
+	)
+	logger.Infof("Starting initialization and precheck for backfilling the database at %s, phase 1 of 4", localDbPath)
+	logger.Infof("Creating and setting LedgerBackend")
+	backend, err := makeBackend(dsInfo)
+	if err != nil {
+		return fmt.Errorf("could not create ledger backend: %w", err)
+	}
+	dsInfo.backend = backend
+	defer backend.Close()
+
+	// Determine what ledgers have been written to local DB
+	// Note GetLedgerRange assumes lexicographical ordering of ledger files in datastore
+	ledgerRange, err := localDbReader.GetLedgerRange(ctx)
+	if err != nil && err != db.ErrEmptyDB {
+		return fmt.Errorf("error getting ledger range from local DB: %w", err)
+	} else if err == db.ErrEmptyDB {
+		dbIsEmpty = true
+	}
+	maxWrittenLedger, minWrittenLedger := ledgerRange.LastLedger.Sequence, max(ledgerRange.FirstLedger.Sequence, 1)
+
+	// Phase 1: precheck to ensure no gaps in local DB
+	if !dbIsEmpty && (maxWrittenLedger >= minWrittenLedger) {
+		logger.Infof("Starting precheck for backfilling the database at %s, phase 1 of 4", localDbPath)
+		if err = verifyDbGapless(ctx, localDbReader, minWrittenLedger, maxWrittenLedger); err != nil {
+			return fmt.Errorf("backfill precheck failed: %w", err)
```

Comment:
in general you should use `errors.Wrap` instead of formatting a new error with the old error inside of it as a string, because the latter method breaks the `errors.Is`ability of a returned value